maven-embedder/src/main/java/org/apache/maven/cli/MavenCli.java

```diff
@@ -1204,40 +1211,57 @@ void toolchains(CliRequest cliRequest) throws Exception {
                         "The specified user toolchains file does not exist: " + userToolchainsFile);
             }
         } else {
-            userToolchainsFile = DEFAULT_USER_TOOLCHAINS_FILE;
+            String userToolchainsFileStr = cliRequest.getUserProperties().getProperty(Constants.MAVEN_USER_TOOLCHAINS);
+            if (userToolchainsFileStr != null) {
+                userToolchainsFile = new File(userToolchainsFileStr);
+            }
         }
 
-        File globalToolchainsFile;
+        File installationToolchainsFile = null;
 
-        if (cliRequest.commandLine.hasOption(CLIManager.ALTERNATE_GLOBAL_TOOLCHAINS)) {
-            globalToolchainsFile =
+        if (cliRequest.commandLine.hasOption(CLIManager.ALTERNATE_INSTALLATION_TOOLCHAINS)) {
+            installationToolchainsFile =
+                    new File(cliRequest.commandLine.getOptionValue(CLIManager.ALTERNATE_INSTALLATION_TOOLCHAINS));
+            installationToolchainsFile = resolveFile(installationToolchainsFile, cliRequest.workingDirectory);
+
+            if (!installationToolchainsFile.isFile()) {
+                throw new FileNotFoundException(
+                        "The specified installation toolchains file does not exist: " + installationToolchainsFile);
+            }
+        } else if (cliRequest.commandLine.hasOption(CLIManager.ALTERNATE_GLOBAL_TOOLCHAINS)) {
+            installationToolchainsFile =
                     new File(cliRequest.commandLine.getOptionValue(CLIManager.ALTERNATE_GLOBAL_TOOLCHAINS));
-            globalToolchainsFile = resolveFile(globalToolchainsFile, cliRequest.workingDirectory);
+            installationToolchainsFile = resolveFile(installationToolchainsFile, cliRequest.workingDirectory);
 
-            if (!globalToolchainsFile.isFile()) {
+            if (!installationToolchainsFile.isFile()) {
                 throw new FileNotFoundException(
-                        "The specified global toolchains file does not exist: " + globalToolchainsFile);
+                        "The specified installation toolchains file does not exist: " + installationToolchainsFile);
             }
         } else {
-            globalToolchainsFile = DEFAULT_GLOBAL_TOOLCHAINS_FILE;
+            String installationToolchainsFileStr =
+                    cliRequest.getUserProperties().getProperty(Constants.MAVEN_INSTALLATION_TOOLCHAINS);
+            if (installationToolchainsFileStr != null) {
+                installationToolchainsFile = new File(installationToolchainsFileStr);
+                installationToolchainsFile = resolveFile(installationToolchainsFile, cliRequest.workingDirectory);
+            }
```

Comment:
This entire block makes me think: WE fail here if the toolchains file cannot be loaded, but we do not fail if core extensions file is provided, but not valid? FNFE and frieds...

Reply:
Or is it `{maven.rootDirectory}`/`{project.rootDirectory}`, anyway it should be consistent with the interpolation style.

Reply:
Then it should be `${session.rootDirectory}`

Reply:
> Then it should be `${session.rootDirectory}`

That'd be perfect.